Pin configuration and builder

bfrog opened the pull request:
Adds a pin configuration struct and set of builder like functions
for it. Easily used with const values or at run time with the
builder pattern if needed. Every pad!() macro call now needs to pass its PAD_CTL_PAD related register in

bfrog commented:
I'd like to extend this to generate const pin configs from settings at compile time as well though I'm not sure if that's possible yet, and what macro magic I'd need to make it happen. probably something like modify_reg!()

bfrog commented:
Ok pretty happy with this now, thinking maybe to re-export each fields options as a module level enum named nicer, but yeah, overall, I think this is a nice way of dealing with the pad_ctl_pad register configuration over the modify_reg! type macros used prior.

bfrog commented on imxrt-hal/src/iomuxc/gpio/b0.rs:
```diff
@@ -1,81 +1,97 @@
 pad!(
     GPIO_B0_00,
     SW_MUX_CTL_PAD_GPIO_B0_00,
+    SW_MUX_CTL_PAD_GPIO_B0_00,
```
Thanks for spotting this copy paste blunder, always the worst

bfrog replied:
Fixed

bfrog commented on imxrt-hal/src/iomuxc/pin_config.rs:
```diff
@@ -0,0 +1,197 @@
+/// the Pad to adjust the PAD_CTL_PAD register in as few operations as possible
+pub struct PinConfig {
+    pub mask: u32,
+    pub value: u32,
```
I was debating that myself, the thing that made me put it pub was the idea that I could, in for example the imxrt1060evk-bsp, define the led function with a const PinConfig value using the macro. I don't believe it'd be possible if this were private.

mciantyre replied:
If we want the best of `const` while maintaining member privacy, could we support a `const fn` as the constructor?

```rust
pub const fn config(
    slew_rate: SlewRate,
    drive_strength: DriveStrength,
    open_drain: OpenDrain,
    /* other fields */
) -> PinConfig {
    /* ... */
}
```

The arguments are enums that could be evaluated at compile time. They're implemented by combining the existing RAL values with the RAL offsets.

```rust
use ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::*;

#[repr(u32)]
pub enum SlewRate {
    Slow = SRE::RW::SRE_0_Slow_Slew_Rate << SRE::RW::offset,
    Fast = SRE::RW::SRE_1_Fast_Slew_Rate << SRE::RW::offset,
}
```

The implementation of `config` just needs to OR the enums together into a `u32`. If we require that users always specify each field (or support a const `default()` function for convenience), we might not need to support the `is_modify()` API. It always becomes a store in the `configure()` method, since users are required to think though every field's values.

bfrog replied:
I wasn't sure to be honest on the const fn constructors, I was under the impression const fn wasn't stabilized yet. If this works, I think its a great idea.

bfrog replied:
I added the const constructor as described here as PinConfig::with_all()

There's also a PinConfig::with_none() which allows incrementally constructing a value using the const fn constructor funcs you described elsewhere. I'm interested to find out how smart const fn is at removing baggage at compile time, though even if it exists still I don't believe it will be very costly for an infrequently changed value.

bfrog commented on imxrt-hal/src/iomuxc/pin_config.rs:
```diff
@@ -0,0 +1,209 @@
+mod slew_rate {
+    use crate::ral;
+
+    pub const MASK: u32 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::mask;
```
It is definitely arbitrary, it happened to be the first iomux register where things are defined for 1062, perhaps not all of them though. I'll find out when I port to the 1010evk

mciantyre replied:
That's a good point; we might need to pick a pad that exists across all systems, or be ready to conditionally compile.

Could we support a private module that does a re-export under an easier-to-type name?

```rust
mod PAD {
    /// All pads support the same configuration; arbitrary choice
    pub user ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::*;
}

pub mod slew_rate {
    // Use pad here --v
    pub const MASK = PAD::DSE::mask;
}
```

bfrog replied:
I like this idea, would make things a lot nicer here, and conditional compiling if needed, simpler. Good idea!

bfrog replied:
Done

bfrog commented on imxrt-hal/src/iomuxc/macros.rs:
```diff
@@ -156,6 +156,21 @@ macro_rules! pad {
             }
+
+            /// Configure pin settings
+            pub fn configure(&mut self, cfg: $crate::iomuxc::pin_config::PinConfig) {
```
This is most definitely meant to be a public API, see the imxrt1060evk-bsp https://github.com/imxrt-rs/imxrt1060evk-bsp/blob/master/src/lib.rs#L221 for where it'd be used. This is based on the zephyr imxrt1060evk pinmux config found here https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/mimxrt1060_evk/pinmux.c

bfrog commented on imxrt-hal/src/iomuxc/pin_config.rs:
```diff
@@ -0,0 +1,218 @@
+    /// Modify an existing given register value with pin config settings
+    pub fn modify(&self, reg: u32) -> u32 {
+        (reg & !self.mask) | (self.value & self.mask)
+    }
```
This I think should be pub(crate), good call.

bfrog commented on imxrt-hal/src/iomuxc/gpio/ad_b0.rs:
```diff
@@ -1,85 +1,101 @@
 pad!(
     GPIO_AD_B0_08,
     SW_MUX_CTL_PAD_GPIO_AD_B0_08,
+    SW_MUX_CTL_PAD_GPIO_AD_B0_08,
```
Thanks for spotting this copy paste blunder, always the worst

bfrog replied:
Fixed

mciantyre commented:
Thanks for the links in your comments! Is the goal to make pad configurations available to a BSP implementer?

Today's HAL hard-codes pad configurations that might not hold across all systems. After this PR, it will be a BSP implementer's responsibility to set the pad configurations for their peripherals. Is that the direction we're going?

mciantyre commented on imxrt-hal/src/iomuxc/pin_config.rs:
```diff
@@ -0,0 +1,218 @@
+    use crate::ral;
+
+    pub const MASK: u32 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::mask;
+    pub const OFFSET: u32 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::offset;
```
Looks like we're using `DSE` here. Should this be `SRE` for slew rate?

Comment also applies to the `MASK` definition.

bfrog replied:
Yes, thank you

bfrog commented:
> Thanks for the links in your comments! Is the goal to make pad configurations available to a BSP implementer?
> 
> Today's HAL hard-codes pad configurations that might not hold across all systems. After this PR, it will be a BSP implementer's responsibility to set the pad configurations for their peripherals. Is that the direction we're going?

I think it makes sense to provide defaults for peripherals in the HAL, this is what mcuxpresso and zephyr both do. On the other hand mcuxpresso also exposes this functionality to application and board specific package code. I think that's the right thing to do. I don't think a HAL should hide access to hardware features, but enable using them in more convenient ways than register access.

One thing here that comes to mind is making better sense of the pull up setup perhaps, rather than directly exposing all the fields provide perhaps an enum that makes more sense. Something I tried to look for clarity on in the ref manual briefly for but didn't see. Might've missed it, the ref manual is huge.

bfrog commented:
No reply next to the const fn idea, but one downside of that I see is that you would need to supply all the fields, where as the pin_config!() macro lets you supply only the fields you need to adjust from the default.

As an example use case...

``` rust
const uart_tx_pin_cfg = pin_config!(pull_keeper_enable: Enabled,
                                                         speed: Speed2_150MHz,
                                                         drive_strength: R0_DIV_6,);
```

Is brief, though perhaps not clear what the other fields might be.

mciantyre commented:
I'm in favor of being explicit, but I can appreciate how concise the Zephyr approach is. It makes sense if the developer knows that the default, zero value is what they want. But, it's not as discoverable; I don't know what I can pass into `IOMUXC_SetPinConfig` without more searching. For instance, I might have to search for other usages of `IOMUXC_SetPinConfig` to learn that I can also specify `IOMUXC_SW_PAD_CTL_PAD_ODE_MASK`. I scrolled down 40 more lines from the linked call site to find that. Requiring all fields lets a new contributor understand what they can supply to the configuration function from a single usage.

If we another `const` approach, [here's a macro-less `const` builder](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=112132641252b006a4b80a1fc5b48aa2) that may guarantee compile-time generation, and lets us elide default value that we don't want to specify.

Could a macro-based approach still incorporate strong types for configuration fields without verbose compiler errors? I'm a fan of strongly-typing the values of the fields. It could help prevent typos, or a user accidentally specifying the wrong value

```rust
const uart_tx_pin_cfg = pin_config!(
    pull_keeper_enable: Enabled,
    speed: FastSlewRate, // should be Speed2_150MHz, not slew rate
    drive_strength: R0_DIV_6,
);
```

mciantyre commented:
> I don't think a HAL should hide access to hardware features, but enable using them in more convenient ways than register access.

I see where this is going. When I hacked together the HAL, I didn't consider that other systems might need different pad configurations. I defaulted to values that were useful for the Teensy 4.

Thanks for putting this together! This will be a great feature for iMXRT users. It could be part of #26, as you noted in that discussion.

If we decide to remove defaults and require HAL users / BSP implementers to always specify their pad configurations, I'm happy to migrate the current configurations from the HAL and re-implement them in the `teensy4-bsp` (or expose a pass-through for BSP users).

bfrog commented:
I'm very happy with this now, I've changed all the builder functions to look like const setters as you had in your example and added a lot of doc comments on what each field is actually doing from the reference manual. I've added the const constructor as desired, with it being the default shown in the docs along with test cases for each default pad_ctl_pad value I could find in the ref manual.

bfrog commented:
Quite happy to merge that one! Thank you for reviewing!
